11_FFTBloom/main.cpp

```diff
@@ -698,74 +672,107 @@ class FFTBloomApp final : public application_templates::MonoDeviceApplication, p
-				imgBarriers[i].image = m_kernelNormalizedSpectrums[i]->getCreationParameters().image.get();
-				imgBarriers[i].subresourceRange.aspectMask = IImage::EAF_COLOR_BIT;
-				imgBarriers[i].subresourceRange.levelCount = 1u;
-				imgBarriers[i].subresourceRange.layerCount = 1u;
-				imgBarriers[i].barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
-				imgBarriers[i].barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
-				imgBarriers[i].barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
-				imgBarriers[i].barrier.dep.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
-				imgBarriers[i].oldLayout = IImage::LAYOUT::UNDEFINED;
-				imgBarriers[i].newLayout = IImage::LAYOUT::GENERAL;
-			}
+			// Also set kernel channel image array to GENERAL for writing
+			decltype(pipelineBarrierInfo)::image_barrier_t imgBarrier = {};
+			pipelineBarrierInfo.imgBarriers = { &imgBarrier, 1 };
 
-			m_computeCmdBuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS(0), pipelineBarrierInfo);
+			imgBarrier.image = m_kernelNormalizedSpectrums->getCreationParameters().image.get();
+			imgBarrier.subresourceRange.aspectMask = IImage::EAF_COLOR_BIT;
+			imgBarrier.subresourceRange.levelCount = 1u;
+			imgBarrier.subresourceRange.layerCount = 3u;
+			imgBarrier.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::NONE;
+			imgBarrier.barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
+			imgBarrier.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			imgBarrier.barrier.dep.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			imgBarrier.oldLayout = IImage::LAYOUT::UNDEFINED;
+			imgBarrier.newLayout = IImage::LAYOUT::GENERAL;
+
+			kernelPrecompCmdBuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS(0), pipelineBarrierInfo);
 
 			//Finally, normalize kernel Image - same number of workgroups
-			m_computeCmdBuf->bindComputePipeline(pipelines[2].get());
-			m_computeCmdBuf->bindDescriptorSets(asset::EPBP_COMPUTE, pipelines[2]->getLayout(), 0, 1, &kernelFFTDescriptorSets[1].get());
-			m_computeCmdBuf->dispatch(paddedKerDim.width / 2, 1, 1);
-			m_computeCmdBuf->end();
+			kernelPrecompCmdBuf->bindComputePipeline(pipelines[2].get());
+			kernelPrecompCmdBuf->dispatch(kerDim.width / 2, 1, 1);
+
+			// Pipeline barrier: transition kernel spectrum images into read only, and outImage into general
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo imagePipelineBarrierInfo = {};
+			decltype(imagePipelineBarrierInfo)::image_barrier_t imgBarriers[2] = {};
+			imagePipelineBarrierInfo.imgBarriers = { imgBarriers, 2 };
+
+			// outImage just needs a layout transition before it can be written to
+			imgBarriers[0].image = m_outImgView->getCreationParameters().image.get();
+			imgBarriers[0].barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::NONE;
+			imgBarriers[0].barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
+			imgBarriers[0].barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::NONE;
+			imgBarriers[0].barrier.dep.dstAccessMask = ACCESS_FLAGS::NONE;
```

Comment:
you need destination masks though, because you want the layout transition to happen before the image gets written

Reply:
or write a comment that the semaphore signal will take care of it as we'll wait on the host before doing any more work